activerecord/lib/active_record/autosave_association.rb

```diff
@@ -401,11 +402,14 @@ def save_collection_association(reflection)
                 if autosave
                   saved = association.insert_record(record, false)
                 elsif !reflection.nested?
+                  association_saved = association.insert_record(record)
+
                   if reflection.validate?
-                    valid = association_valid?(reflection, record, index)
-                    saved = valid ? association.insert_record(record, false) : false
-                  else
-                    association.insert_record(record)
+                    if !association_saved
+                      association_valid?(reflection, record, nil, false)
```

Comment:
How about just using `errors.add(reflection.name)`?
The `record` in this context is, not destroyed, not `marked_for_destruction?`, not `reflection.options[:autosave]` given.

```diff
diff --git a/activerecord/lib/active_record/autosave_association.rb b/activerecord/lib/active_record/autosave_association.rb
index a405f05e0b..ed71e68ad1 100644
--- a/activerecord/lib/active_record/autosave_association.rb
+++ b/activerecord/lib/active_record/autosave_association.rb
@@ -392,7 +392,7 @@ def save_collection_association(reflection)
               records -= records_to_destroy
             end
 
-            records.each_with_index do |record, index|
+            records.each do |record|
               next if record.destroyed?
 
               saved = true
@@ -402,8 +402,8 @@ def save_collection_association(reflection)
                   saved = association.insert_record(record, false)
                 elsif !reflection.nested?
                   if reflection.validate?
-                    valid = association_valid?(reflection, record, index)
-                    saved = valid ? association.insert_record(record, false) : false
+                    saved = association.insert_record(record)
+                    errors.add(reflection.name) unless saved
                   else
                     association.insert_record(record)
                   end
```